src/libraries/SafeCast.sol

```diff
@@ -6,35 +6,55 @@ pragma solidity ^0.8.20;
 library SafeCast {
     error SafeCastOverflow();
 
+    function _revertOverflow() private pure {
+        /// @solidity memory-safe-assembly
+        assembly {
+            // Store the function selector of `SafeCastOverflow()`.
+            mstore(0x00, 0x93dafdf1)
+            // Revert with (offset, size).
+            revert(0x1c, 0x04)
+        }
+    }
+
     /// @notice Cast a uint256 to a uint160, revert on overflow
-    /// @param y The uint256 to be downcasted
-    /// @return z The downcasted integer, now type uint160
-    function toUint160(uint256 y) internal pure returns (uint160 z) {
-        z = uint160(y);
-        if (z != y) revert SafeCastOverflow();
+    /// @param x The uint256 to be downcasted
+    /// @return The downcasted integer, now type uint160
+    function toUint160(uint256 x) internal pure returns (uint160) {
+        if (x >= 1 << 160) _revertOverflow();
```

Comment:
technically 1<<160 and 1 << 128 can be constants which may make this even cheaper?

Reply:
The compiler would compute it to a constant at compile time. It increases readability imo.